crates/core/src/subscription/execution_unit.rs

```diff
@@ -132,24 +183,38 @@ impl ExecutionUnit {
         tables: impl Iterator<Item = &'a DatabaseTableUpdate>,
         auth: AuthCtx,
     ) -> Result<Option<DatabaseTableUpdate>, DBError> {
-        let ops = match self.plan.kind {
-            Supported::Select => {
-                let mut ops = Vec::new();
-                for table in tables.filter(|table| table.table_id == self.return_table()) {
-                    // Replace table reference in original query plan with virtual MemTable
-                    let (plan, sources) = query::to_mem_table(self.plan.expr.clone(), table);
-                    // Evaluate the new plan and capture the new row operations.
-                    ops.extend(eval_primary_updates(db, auth, tx, &plan, sources)?.map(|r| TableOp::new(r.0, r.1)));
-                }
-                ops
+        let ops = if let Some(eval_incr_plan) = &self.eval_incr_plan {
+            debug_assert!(matches!(self.eval_plan.kind, Supported::Select));
+            let SourceExpr::MemTable {
+                source_id: _source_id,
+                ref header,
+                table_access,
+                ..
+            } = eval_incr_plan.source
+            else {
+                panic!("Expected MemTable in `eval_incr_plan`, but found `DbTable`");
+            };
+            let mut ops = Vec::new();
+            for table in tables.filter(|table| table.table_id == self.return_table()) {
+                // Build a `SourceSet` containing the updates from `table`.
+                let mem_table = query::to_mem_table_with_op_type(header.clone(), table_access, table);
+                let mut sources = SourceSet::default();
+                let _source_expr = sources.add_mem_table(mem_table);
+                debug_assert_eq!(_source_expr.source_id(), Some(_source_id));
+                // Evaluate the saved plan against the new `SourceSet`
+                // and capture the new row operations.
+                ops.extend(
+                    eval_primary_updates(db, auth, tx, eval_incr_plan, sources)?.map(|r| TableOp::new(r.0, r.1)),
```

Comment:
Ideally `eval_primary_updates` would just instantiate the iterator and run through the results. But right now it creates another intermediate representation `QueryCode`, and that's what is passed to `build_query`.

So to avoid cloning the QueryExpr, every time we call `eval_primary_updates`, we'll probably want to have it operate directly on `QueryCode`.

And I think ultimately this means that an ExecutionUnit should hold a QueryCode rather than a QueryExpr.

Reply:
I changed `ExecutionUnit` so that, for `Select` queries, it stores `QueryCode` instead of `QueryExpr`. This does not remove the clones, as `build_query` takes an owned `QueryCode`. We could change that pretty easily, I think, but I'd like it to wait for a follow-up PR.